usr/lib/common/pkey_utils.c

```diff
@@ -31,6 +31,72 @@
 #include "trace.h"
 #include "pkey_utils.h"
 
+#if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+# if __GLIBC_PREREQ(2, 16)
+#  include <sys/auxv.h>
+#  if defined(HWCAP_S390_STFLE) && defined(HWCAP_S390_VX)
+#   define IMPLEMENT_GETAUXVAL
+#  endif
+# endif
+#endif
+
+/**
+ * s390_stfle:
+ *
+ * Executes the STFLE operation of the CPU.
+ *
+ * Returns the no. of double words needed to store the facility bits.
+ */
+static inline int s390_stfle(unsigned long long *list, int doublewords)
+{
+    register unsigned long __nr asm("0") = doublewords - 1;
+
+    asm volatile(".insn s,0xb2b00000,0(%1)" /* stfle */
+             : "+d" (__nr) : "a" (list) : "memory", "cc");
+
+    return __nr + 1;
+}
+
+/**
+ * Determine the machine's MSA level.
+ */
+int get_msa_level(void)
+{
+    unsigned long long facility_bits[3];
+    int msa = 0;
+    int num = 0;
+#ifdef IMPLEMENT_GETAUXVAL
+    const unsigned long hwcap = getauxval(AT_HWCAP);
+
+    if (hwcap & HWCAP_S390_STFLE) {
+#endif
+        memset(facility_bits, 0, sizeof(facility_bits));
+        num = s390_stfle(facility_bits, 3);
```

Comment:
> so in case IMPLEMENT_GETAUXVAL is undefined s390_stfle is still called ?

Yes, this was my intention. If getauxval is undefined, because we are running on an old kernel, STFLE could probably still be available, because we are running on new hardware. However, yes, if running on old hw or on an emulator not supporting STFLE, this would crash. Do we want to be defensive here?

Reply:
Yes. Id say its okay to not support this feature on systems that dont have getauxval

Reply:
Done.